decavsub: add general purpose avcodec subtitle decoder

jstebbins opened the pull request:
Currently using it for pgs, srt, and ssa subtitles.

Fixes srt font face parsing.

It's fairly easy to add more or convert our current decoders to libav.  I'll look into converting our vobsub and cc608 subtitle decoders to use libav as well.  And adding dvbsubs should be relatively easy.

**Test on:**

- [x] Windows 10+  (via MinGW)
- [x] macOS 10.13+
- [x] Ubuntu Linux

jstebbins committed:
decavsub: add general purpose avcodec subtitle decoder

Currently using it for pgs, srt, and ssa subtitles.

jstebbins committed:
decavsub: add DVB subtitle *burn-only* support

passthrough will hopefully come later ;)

jstebbins committed:
decavsub: enable dvb subtitle passthrough in mkv

jstebbins committed:
decavsub: use libav to decode EIA 608 subtitles

simplifies code, eliminates deccc608sub.c

sr55 commented:
Related issues: 
#982 DVB Subtitles
#1799 CC608 / 708
#2494 SRT Subtitle font 
#2449 Subtitle Rendering issues

jstebbins committed:
decavsub: use libav to decode dvd subtitles

Simplifies code, removes encvobsub.c (was never used) and decvobsub.c.

jstebbins committed:
decavsub: remove dvb subtitle post-stuffing during demux

This seems to be what ffmpeg expects.  This way we also mux it correctly
into mkv, i.e. without the trailing 0xff stuffing byte.

jstebbins committed:
decavsub: fix dvb subtitle passthrough

The "clear" subtitle packets were not making it through to the muxer

jstebbins committed:
fix job logging of dvb subtitle source

sr55 commented:
Some preliminary testing on Windows:

| Subtitle Type | Test | Tested |
| ------------- | ------------- | ------------- |
| VobSub  | Passthru | Pass |
| VobSub  | Burn-in | Pass |
| ASS | Passthru | Pass |
| ASS | Burn-in | Pass |
| ASS | Burn-in + Offset | Pass |
| SRT |  Passthru  | Pass |
| SRT |  Burn-in | Pass (Including Bold, Italitc etc) |
| CC608 |  Passthru  | Pass |
| CC608 |  Burn-in | Odd Rendering (Screenshots below) |
| PGS | Passthru | Pass  |
| PGS | Burn-in | Pass |

![image](https://user-images.githubusercontent.com/628593/70850557-92c2a400-1e83-11ea-8be3-be8f8e11159e.png)
 vs passthru:

![image](https://user-images.githubusercontent.com/628593/70850564-ae2daf00-1e83-11ea-819f-f20e80fa8c90.png)

jstebbins commented:
Nice! Thanks for all the testing.  I think I have that particular sample you see the odd rendering with (I recognize that desert scene :wink: )  I'm seeing broken timing of the subtitles with that sample, will look into that.  The rendering is almost the same with the player I use (mpv)  so the font size difference must be your player making an odd choice.  The the position is too close to the bottom and subtitles are being cropped when burned in.

sr55 commented:
It's VLC and yes, it's the Doctor who sample :)

jstebbins committed:
decavsub: reduce EIA 608 font size

The larger font size was causing rendering outside the frame boundaries

jstebbins commented:
I get identical results between burned, mpv, and vlc.  Could be a difference in font selection on windows.  On linux CC font is "monospace" which resolves to "DejaVuSansMono" on my system. And on windows CC uses "Lucida Console".  We chose that a while back for windows because monospace resulted in selection of "Courier New" which looked awful. Maybe Lucida Console has limited range of font sizes?  The definition of this is in handbrake/internal.h HB_FONT_MONO if you want to try other fonts.

jstebbins committed:
decavsub: patch ffmpeg to fix workaround hacks

sr55 commented:
20 is better.

Maybe something we can make configurable down the line.

sr55 commented:
Shall we push this and let folks have a play?

jstebbins commented:
Be sure to `make ffmpeg.xclean` if you are testing from an existing build.  There are patches that must be applied.

jstebbins commented:
FYI, the features this adds are
* More complete handling of SRT formatting tags
* DVB subtitle burn-in and "passthrough" to mkv.  mp4 must be burned (bitmap subs)

cehoyos commented:
Please send your patches upstream for review.

cehoyos commented:
And while there, please look at FFmpeg tickets 7377, 7389, 7572, 8070 and 8502

jstebbins commented:
@cehoyos fix for mp4 issues posted
https://patchwork.ffmpeg.org/project/ffmpeg/patch/20200329203245.225679-1-jstebbins@jetheaddev.com/
